network/index.md

```diff
@@ -1,103 +1,137 @@
-
-- [Standalone networking tutorial](network-tutorial-standalone.md)
-- [Host networking tutorial](network-tutorial-host.md)
-- [Overlay networking tutorial](network-tutorial-overlay.md)
-- [Macvlan networking tutorial](network-tutorial-macvlan.md)
+Container networking refers to the ability for containers to connect to and
+communicate with each other, or to non-Docker workloads.
+
+A container has no information about what kind of network it's attached to,
+or whether their peers are also Docker workloads or not.
+A container only sees a network interface with an IP address,
+a gateway, a routing table, DNS services, and other networking details.
+That is, unless the container uses the `none` network driver.
+
+This page describes networking from the point of view of the container,
+and the concepts around container networking.
+This page doesn't describe OS-specific details about how Docker networks work.
+For information about how Docker manipulates `iptables` rules on Linux,
+see [Packet filtering and firewalls](packet-filtering-firewalls.md).
+
+## Published ports
+
+By default, when you create or run a container using `docker create` or `docker run`,
+the container doesn't expose any of its ports to the outside world.
+Use the `--publish` or `-p` flag to make a port available to services
+outside of Docker.
+This creates a firewall rule in the host,
+mapping a container port to a port on the Docker host to the outside world.
+Here are some examples:
+
+| Flag value                      | Description                                                                                                                                           |
+| ------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `-p 8080:80`                    | Map TCP port 80 in the container to port `8080` on the Docker host.                                                                                   |
+| `-p 192.168.1.100:8080:80`      | Map TCP port 80 in the container to port `8080` on the Docker host for connections to host IP `192.168.1.100`.                                        |
+| `-p 8080:80/udp`                | Map UDP port 80 in the container to port `8080` on the Docker host.                                                                                   |
+| `-p 8080:80/tcp -p 8080:80/udp` | Map TCP port 80 in the container to TCP port `8080` on the Docker host, and map UDP port `80` in the container to UDP port `8080` on the Docker host. |
+
+> **Important**
+>
+> Publishing container ports is insecure by default. Meaning, when you publish
+> a container's ports it becomes available not only to the Docker host, but to
+> the outside world as well.
+>
+> If you include the localhost IP address (`127.0.0.1`) with the publish flag,
+> only the Docker host, and other systems connected to the same LAN, can access
```

Comment:
I haven't had time to look at everything yet, but I didn't want to potentially let this slip by:

LAN isn't accurate here; a LAN is just that -- a local network, which can be switched or routed.

This is referring to the ability to use a container host as a gateway that routes packets to the container bridge subnet by adding an explicit route: this **only** works on the same switched (and not routed) network segment.

Reply:
That'd be my fault as I don't know networking that well and only understood earlier discussions referred to [other hosts on a LAN](https://github.com/moby/moby/issues/22054#issuecomment-1163696666) (_and something about network supporting [layer 2 switching](https://github.com/moby/moby/issues/14041#issuecomment-223467379)?_).

I could easily reproduce it with two VMs in the same network, and two VPS instances in the same network. Connecting to ports published on the docker host at `127.0.0.1` and without firewalld to docker bridges (eg: `172.17.0.0/16`).

I [suggested that the earlier iteration may want to also warn users](https://github.com/docker/docs/pull/17176#discussion_r1212620404) about connectivity not being only restricted to `127.0.0.1` within the Docker host, as one would have expected, but wasn't sure how to communicate that.

I know that it's not reachable by default, and requires the other host to configure an explicit route to the docker hosts target subnet. That'd probably get a bit verbose to detail in these docs though?

---

For the docs perhaps warning of an exception for `-p 127.0.0.1:host_port:container_port` that links to the [associated Github issue](https://github.com/moby/moby/issues/45610) would be appropriate?

Reply:
I think tweaking the wording here is fine, unless we become certain this will be the status quo for a long time. I'm hoping that if we can find a way to keep those packets out without breaking any other aspect of behavior, we can just call it a bug and backport the change.

In any case, "supporting layer 2 switching" is something all networks do -- basically, for this to work you have to not cross a router (as otherwise it's up to **the router's routing table** where the packets go, and not solely your local machine). You must be on the **same layer 2 fabric* -- for a home network, that's the same thing as the LAN.

However, the moment you introduce additional routed local networks that collectively make a local **area** network (e.g. multiple VLANs, each with their own subnet, with a router moving packets between them), this equivalence breaks down; thus the need to be more precise here.

Reply:
> I'm hoping that if we can find a way to keep those packets out without breaking any other aspect of behavior, we can just call it a bug and backport the change.

For `127.0.0.1` / `127.0.0.0/8` I think that should be fine with the [added restriction on `PREROUTING`](https://github.com/moby/moby/issues/45610#issuecomment-1562191303). I don't know about the other issue with routing to docker networks, linking to the issue for more information at least provides a workaround and can be subscribed for updates/fix. I assume the upcoming IPVS alternative might also avoid it, if so affected users could be encouraged to migrate to that instead.

> thus the need to be more precise here.

You probably understand the docs audience better than I would.

The bulk of the users that actually read the docs may lack the expertise for the extra precision in docs to be of value to them, thus the added verbosity may be better served by linking to the dedicated Github issue?

- It'd keep the docs more focused by reducing noise, while providing all the technical details to those curious or with the relevant expertise to make sense of it (_and potentially provide positive contributions_).
- It's also content that ideally isn't relevant to expand on in the docs since it's more of a bug than actual supported functionality to document.
- Even for those that don't grok the details well, a tracking issue at least provides visibility of the status (_as docs may fall out of sync, like this PR drops outdated information regarding forwarding that was reported years ago_), plus a subscribe button to be notified when it's resolved.

---

This suggestion probably needs to be reworded, but might be succinct enough?

```suggestion
> only the Docker host can access the published container port.
> **Warning:** Hosts within the same local network may be capable
> of routing to that private network port of the Docker host.
> For details, see [moby/moby#45610](https://github.com/moby/moby/issues/45610).
```

- I'm not sure if you'd refer to it as a private or internal network of the docker host.
- Is routing correct?
  
  I know that it's handled via `ip route` command which directs the IP or subnet specified traffic through to the docker host LAN IP.
  I assume that would otherwise be directed to an upstream router from the default routes gateway? And since the Docker host enabled `ip_forward=1`, it's behaving as a router as you mentioned earlier.

Reply:
Although we discussed it during our daily meeting, I'm still quite on the fence here. I don't feel comfortable to talk about a security issue in our docs, but on the other hand the initial claim was a lie (that ports published on 127.0.0.1 are only accessible from the Docker host), so...

@dvdksn Can you apply this suggestion, or something along the lines (it probably needs to be properly reworded)?

```suggestion
> only the Docker host can access the published container port.
> **Warning:** Hosts within the same L2 segment (eg. connected
> to the same network switch) can reach those ports published to
> 127.0.0.1. For details, see [moby/moby#45610](https://github.com/moby/moby/issues/45610).
```

As for the the other, similar security issues, I think we're ok not talking about them.